### PR TITLE
8271340: Crash PhaseIdealLoop::clone_outer_loop

### PR DESCRIPTION
**Pull request by rwestrel**
The test has a counted loop and an infinite loop. The infinite loop is
reachable from multiple paths and its loop head is a Region with more
than 3 inputs. One of these paths is from the counted loop. When loop
opts run, a NeverBranch is added to the infinite loop that's removed
by NeverBranchNode::Ideal() next because in(0) of the NeverBranch is a
Region and not a Loop.

When CCP runs, it finds the counted loop exit is never reached because
a test in the loop body that depends on a loop phi is never taken. As
a consequence nodes along the path from the counted loop end to the
infinite loop Region have type top. One of these nodes is a Call
node. PhaseCCP::do_transform() should then cause the path between the
counted loop and the infinite loop to optimize out but that doesn't
happen because do_transform() starts from Root by following inputs.
The dead path is only reachable from the infinite loop but the there's
no edge between Root and the infinite loop.

IGVN next runs, processes the Call Node, finds it's dead, kills
everything around it which causes the OuterStripMinedLoopEnd to loose
a projection. That later triggers the crash.

The fix I propose is to be more conservative in
NeverBranchNode::Ideal() and to check for a in(0) that's a Region. As
a consequence, at CCP time, the infinite loop is reachable from Root.

This change also requires some adjustments to Shenandoah specific code
that makes assumptions about the shape of infinite loops.

/cc hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271340](https://bugs.openjdk.java.net/browse/JDK-8271340): Crash PhaseIdealLoop::clone_outer_loop

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 0ece249f52929ad77810929c4257caea0ca4a05c
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5220/head:pull/5220` \
`$ git checkout pull/5220`

Update a local copy of the PR: \
`$ git checkout pull/5220` \
`$ git pull https://git.openjdk.java.net/jdk pull/5220/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5220`

View PR using the GUI difftool: \
`$ git pr show -t 5220`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5220.diff">https://git.openjdk.java.net/jdk/pull/5220.diff</a>

</details>

### CODE REVIEW

**Comment by rwestrel** [comment]
Anyone for this?

File: src/hotspot/share/opto/cfgnode.cpp [hunk 1]
```diff
@@ -2717,7 +2717,7 @@ const Type* NeverBranchNode::Value(PhaseGVN* phase) const {
 //------------------------------Ideal------------------------------------------
 // Check for no longer being part of a loop
 Node *NeverBranchNode::Ideal(PhaseGVN *phase, bool can_reshape) {
-  if (can_reshape && !in(0)->is_Loop()) {
+  if (can_reshape && !in(0)->is_Region()) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/cfgnode.cpp:2720]: I assume this is the actual fix.

**Review by vnkozlov — COMMENTED** [review]
Please, place `TestInfiniteLoopCCP.java` test into `compiler/loopopts`

File: src/hotspot/share/opto/cfgnode.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/cfgnode.cpp:2720]: Yes, it is. The rest is shenandoah specific.

File: test/hotspot/jtreg/compiler/c2/TestInfiniteLoopCCP.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                    TestInfiniteLoopCCP instance = new TestInfiniteLoopCCP();
+                    for (int i = 0; i < 100; ++i) {
+                        instance.mainTest(args);
+                    }
+                }
+            };
+        // Give thread some time to trigger compilation
+        thread.setDaemon(true);
+        thread.start();
+        Thread.sleep(/*Utils.adjustTimeout(4000)*/4000);
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestInfiniteLoopCCP.java:72]: Why is the `Utils.adjustTimeout` uncommented?

**Review by TobiHartmann — COMMENTED** [review]
> but that doesn't happen because do_transform() starts from Root by following inputs

Would the following code I added to Valhalla help?

https://github.com/openjdk/valhalla/commit/f43fafc7d81f3d34f6e971765dbadae41ae5c393#diff-f4a11a2c3f7d7342641c878277f778856ad329e4e09026e39d08afb03efd10a2R1958

File: src/hotspot/share/opto/cfgnode.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/cfgnode.cpp:2720]: okay

**Review by vnkozlov — APPROVED** [review]
Good.
But, please, address Tobias's question about test.

**Comment by rwestrel** [comment]
> Please, place `TestInfiniteLoopCCP.java` test into `compiler/loopopts`

fixed in the update change

File: test/hotspot/jtreg/compiler/c2/TestInfiniteLoopCCP.java (hunk 2 above)
Comment (rwestrel) [reply test/hotspot/jtreg/compiler/c2/TestInfiniteLoopCCP.java]: That was an accident. Fixed in updated change.

**Comment by rwestrel** [comment]
> Would the following code I added to Valhalla help?
> 
> [openjdk/valhalla@f43fafc#diff-f4a11a2c3f7d7342641c878277f778856ad329e4e09026e39d08afb03efd10a2R1958](https://github.com/openjdk/valhalla/commit/f43fafc7d81f3d34f6e971765dbadae41ae5c393#diff-f4a11a2c3f7d7342641c878277f778856ad329e4e09026e39d08afb03efd10a2R1958)

I'm not sure. How would it help in this particular case?

In any case, it seems to me NeverBranchNode::Ideal() is simply wrong. As I understand, when loop opts hit an infinite loop it creates a NeverBranchNode but because the infinite loop is not properly registered in the loop tree, the loop's Region is not transformed into a Loop node. On the next pass of loop opts, if the NeverBranch is still in the infinite loop, a Loop is created for the infinite loop. But if NeverBranchNode::Ideal() runs in between then a new NeverBranch is added and maybe and only in some later loop opts pass a Loop created. So AFAICT, the behavior is inconsistent depending on whether NeverBranchNode::Ideal() runs or not which doesn't seem right.

**Review by TobiHartmann — APPROVED** [review]

**Comment by TobiHartmann** [comment]
> > Would the following code I added to Valhalla help?
> > [openjdk/valhalla@f43fafc#diff-f4a11a2c3f7d7342641c878277f778856ad329e4e09026e39d08afb03efd10a2R1958](https://github.com/openjdk/valhalla/commit/f43fafc7d81f3d34f6e971765dbadae41ae5c393#diff-f4a11a2c3f7d7342641c878277f778856ad329e4e09026e39d08afb03efd10a2R1958)
> 
> I'm not sure. How would it help in this particular case?

I thought it might help by aggressively removing the parts not reachable from the bottom (but still reachable from root).

> In any case, it seems to me NeverBranchNode::Ideal() is simply wrong. As I understand, when loop opts hit an infinite loop it creates a NeverBranchNode but because the infinite loop is not properly registered in the loop tree, the loop's Region is not transformed into a Loop node. On the next pass of loop opts, if the NeverBranch is still in the infinite loop, a Loop is created for the infinite loop. But if NeverBranchNode::Ideal() runs in between then a new NeverBranch is added and maybe and only in some later loop opts pass a Loop created. So AFAICT, the behavior is inconsistent depending on whether NeverBranchNode::Ideal() runs or not which doesn't seem right.

Thanks for the details. That makes sense.

**Comment by rwestrel** [comment]
Thanks for the reviews @vnkozlov @TobiHartmann

**Pull request closed without merging**